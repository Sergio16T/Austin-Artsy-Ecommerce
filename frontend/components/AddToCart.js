import React, { Component } from 'react';
import { Mutation } from 'react-apollo'; 
import gql from 'graphql-tag'; 
import { CURRENT_USER_QUERY } from './User'; 

const ADD_TO_CART_MUTATION = gql`
    mutation ADD_TO_CART_MUTATION($id: ID! ) {
        addToCart(id: $id) {
            id
            quantity
        }
    }
`; 

class AddToCart extends Component {
    render() {
        return (
            <Mutation 
            mutation={ADD_TO_CART_MUTATION} 
            variables={{id: this.props.id}}
            refetchQueries={[{query: CURRENT_USER_QUERY }]}
            >
                {(addToCart, {error, loading}) => 
                    <button type="button" onClick={async () => {
                        const res = await addToCart().catch(err => {
                            alert(err.message); 
                        }); 
                        console.log(res);
                    }}>Add to Cart</button>
                }
            </Mutation>
        );
    }
}

export default AddToCart;