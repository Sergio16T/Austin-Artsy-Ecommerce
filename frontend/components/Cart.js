import React, { Component } from 'react';
import styled from 'styled-components';

const CartWrapper = styled.div`
    position: fixed; 
    height: 100%; 
    width: 400px; 
    transform: ${props => props.isOpen ? "translateX(0)" : "translateX(100%)" }; 
    background: ${props => props.theme.offWhite}; 
    right: 0; 
    top: 0; 
    z-index: 5; 
    border-left: 2px solid ${props => props.theme.purple}; 
    min-width: 500px; 
    transition: all 0.3s;
    overflow: hidden; 
`; 

class Cart extends Component {
    render() {
        return (
            <CartWrapper isOpen={this.props.isOpen}>
                <h2>Cart</h2>
            </CartWrapper>
        );
    }
}

export default Cart;