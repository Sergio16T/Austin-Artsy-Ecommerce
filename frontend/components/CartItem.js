import React from 'react';
import { useMutation } from 'react-apollo'; 
import gql from 'graphql-tag'; 
import styled from 'styled-components';
import RemoveFromCart from './RemoveFromCart'; 
import formatMoney from "../lib/formatMoney";
import { CURRENT_USER_QUERY } from './User';

const UPDATE_CART_ITEM_MUTATION = gql`
    mutation UPDATE_CART_ITEM_MUTATION($updateType: String!, $id: ID!) {
        updateCartItem(updateType: $updateType, id: $id) {
            id
            quantity
        }
    }
`; 

const StyledCartItem = styled.div`
display: grid; 
grid-template-columns: 1fr 2fr .5fr; 
padding: 2rem 0; 
.cartItemImage {
    width: 120px; 
}
@media (max-width: 600px) {
    padding: 1rem 0; 
    .cartItemImage {
        width: 80px; 
    }
}
`; 

const CartItemDescription = styled.div`
    padding-left: 2rem;
.cartitem_title{
    margin-top: 0; 
    font-size: 1.2rem;
}
.priceandquantity {
    display: flex; 
    flex-direction: row; 
}
`; 
const QuantityDiv = styled.div`
    display: flex;
    flex-direction: row; 
    font-size: 1.6rem; 
    padding-left: 2rem; 
    align-items: center; 
.cartItem_Quantity, .minus, .plus {
    padding:  0 .75rem; 
    margin: 0; 
}
.minus, .plus {
    outline: none; 
    border: none; 
    background: transparent; 
    font-size: 1.6rem; 
    cursor: pointer; 
}
.cartitem_price {
    font-size: 1.2rem; 
}
`; 
const CartItem = (props) => {
    const [updateCartItem, { data }] = useMutation(UPDATE_CART_ITEM_MUTATION); 
	const { item } = props.cartItem; 
	return (
		<StyledCartItem>
			<img className="cartItemImage" src={item.image}/> 
			<CartItemDescription>
				<p className="cartitem_title">{item.title}</p>
				<div className="priceandquantity">
					<p className="cartitem_price">{formatMoney(item.price)}</p>
					<QuantityDiv>
						<button className="minus" type="button" onClick={() => {
                            updateCartItem({variables: { updateType: "subtract", id: props.cartItem.id}}); 
                        }} disabled={props.cartItem.quantity === 1}>-</button>
						<p className="cartItem_Quantity">{props.cartItem.quantity}</p>
						<button className="plus" onClick={() => {
                            updateCartItem({variables: { updateType: "add", id: props.cartItem.id}}); 
                        }}>+</button>
					</QuantityDiv>
				</div>
			</CartItemDescription>
			
			<RemoveFromCart id={props.cartItem.id}/>
		</StyledCartItem>
	)
}

export default CartItem; 