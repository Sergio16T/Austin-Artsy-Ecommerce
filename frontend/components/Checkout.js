import React, { Component } from 'react';
import StripeCheckout from 'react-stripe-checkout'; 
import { Mutation } from 'react-apollo'; 
import Router from 'next/router'; 
import Nprogress from 'nprogress'; 
import PropTypes from 'prop-types'; 
import gql from 'graphql-tag'; 
import calculateTotal from '../lib/calculateTotal'; 
import { CURRENT_USER_QUERY } from './User'; 

const CREATE_ORDER_MUTATION = gql`
    mutation CREATE_ORDER_MUTATION($token: String!) {
        createOrder(token: $token) {
            id
						total 
						charge
						items {
							id
							title
						}
        }
    }
`; 

function totalItems(cart) {
    return cart.reduce((startValue, element) => {
        return startValue + element.quantity; 
    }, 0); 
}

class Checkout extends Component {
    onToken = async (res, createOrder) => {
				console.log(res.id); 
				this.props.toggleCart(false);
				const order = await createOrder({
					variables: { token: res.id }
				}).catch(err => {
					console.log(err.message);  
				}); 
				console.log(order); 
				Router.push({
					pathname: "/order", 
					query: {id: order.data.createOrder.id}
				});
			
    }
    render() {
        return (
						<Mutation 
						mutation={CREATE_ORDER_MUTATION}
						refetchQueries={[{ query: CURRENT_USER_QUERY }]}
						>
                {(createOrder, {error, loading}) => 
							<StripeCheckout
							amount={this.props.totalAmount}
							name="Austin's List"
							description={`Order of ${totalItems(this.props.cart)} items`}
							image={this.props.cart.length && this.props.cart[0].item ? this.props.cart[0].item.image : null}
							stripeKey="pk_test_lUHGmLxt0KGqdsIN20tuXiVw00Q2OwgjiR"
							currency="USD"
							email={this.props.email}
							token={(res) => this.onToken(res, createOrder)}
							>{this.props.children}</StripeCheckout>
            }
            </Mutation>
        );
    }
}

export default Checkout;