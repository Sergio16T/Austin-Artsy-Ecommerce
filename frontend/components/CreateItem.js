import React, { Component } from 'react';
import { Query, Mutation } from 'react-apollo'; 
import Router from 'next/router'; 
import gql from'graphql-tag'; 
import { StyledFormWrapper, StyledForm} from './styles/FormStyles'; 
import Spinner from './Spinner'; 
import { ALL_ITEMS_QUERY } from './Items'; 
import { PAGINATION_QUERY } from './Pagination'; 

const CREATE_ITEM_MUTATION = gql`
    mutation CREATE_ITEM_MUTATION(
        $title: String! 
        $description: String!
        $price: Int! 
        $image: String!
        $largeImage: String!
        ) {
        createItem(
            title: $title, 
            description: $description, 
            price: $price
            image: $image
            largeImage: $largeImage 
            ) {
            id
        }
    }
`; 
class CreateItem extends Component {
    state = {
        title: '', 
        price: 0, 
        description: '',
        image: '',
        largeImage: '',
        errorMessage: '',
        uploadError: ''
    }
    uploadFile = async (e) => {
        console.log('uploading file'); 
        this.setState({ spinner: true }); 
        const files = e.target.files; 
        const data = new FormData(); 
        data.append('file', files[0]); 
        data.append('upload_preset', 'AustinArts'); 

        const res = await fetch('https://api.cloudinary.com/v1_1/dddnhychw/image/upload', {
            method: 'POST', 
            body: data  
        }); 
        const file = await res.json(); 
        console.log(file); 

        try {
            this.setState({
                image: file.secure_url, 
                largeImage: file.eager[0].secure_url,
                spinner: false
            }); 
        }
        catch(err) {
            this.setState({uploadError: err.message}); 
        }
    }
    handleInput = (e) => {
        let regEx = /^\d+(\.\d{0,2})?$/; 
        const { value, type, name } = e.target; 
        let val = type === 'number' && value.length ? parseFloat(value) : value;  
        // console.log(regEx.test(val)); 
        try {
            if ((typeof val === 'number') && !regEx.test(val)) {
                throw new Error('Only two decimal points please'); 
            }
        }
        catch(err) {
            console.log(err.message);
            this.setState({
                errorMessage: err.message
            }); 
        }
        if ((typeof val === 'number') && regEx.test(val)) {
            this.setState({
                errorMessage: ''
            }); 
        }
        this.setState({
            [name]: val
        }); 
    }
    render() {
        let regEx2= /^\d+(\.\d{0,1})$/; 
        const { errorMessage } = this.state; 
        const setPrice = regEx2.test(this.state.price) ? this.state.price.toFixed(2) : this.state.price; 
        return (
            <StyledFormWrapper>
                <div className="formContainer">
                    <Query query={PAGINATION_QUERY}>
                        {({data, error, loading}) => {
                            if(loading) return null; 
                            const count = data.itemsConnection.aggregate.count; 
                            // console.log(count); 
                            let lastOnPage  = count/4 === 0; 
                            let page = Math.ceil((count+1)/4) 
                            // console.log(page); 
                            let skip = page * 4 - 4;
                            // console.log(skip); 
                            return (
                        <Mutation 
                        mutation={CREATE_ITEM_MUTATION} 
                        variables={{...this.state, price: setPrice * 100}}
                        refetchQueries={[
                            { query: PAGINATION_QUERY }, 
                            { query: ALL_ITEMS_QUERY, variables: {skip: skip} }, 
                            { query: ALL_ITEMS_QUERY, variables: {skip: (page-1) * 4 - 4 }}
                        ]}
                        >
                            {(createItem, {loading, error}) => (
                            <StyledForm onSubmit={ async (e)=> {
                                e.preventDefault(); 
                                this.setState({ spinner: true });
                                let res = await createItem().catch(err => {
                                    if(err) {
                                        this.setState({ 
                                            spinner: false, 
                                            error: true
                                         }); 
                                    } else {
                                        this.setState({
                                            error: false
                                        }); 
                                    }  
                                }); 
                                this.setState({ spinner: false }); 
                                console.log(res); 
                                if(!this.state.error) {
                                    Router.push({
                                        pathname: "/items", 
                                        query: {page: page}
                                        // query: {id: res.data.createItem.id}
                                    }); 
                                }
                            }} 
                            errorMessage = {errorMessage}
                            >
                                <Spinner spinner={this.state.spinner}/>
                                {error && <p className="errorMessage">{error.message.replace("GraphQL error:", "")}</p>}
                                <fieldset disabled={loading} aria-busy={loading}>
                                    <label htmlFor="file">
                                        Image
                                        <input
                                        type ="file" 
                                        id ="file" 
                                        name="file" 
                                        placeholder="upload an image" 
                                        onChange={this.uploadFile}
                                        />
                                    </label>
                                    {this.state.largeImage && <img width="250" src={this.state.image} alt={this.state.description}></img>}
                                    <label htmlFor="title">
                                        Title
                                        <input
                                        type ="text" 
                                        id ="title" 
                                        name="title" 
                                        placeholder="title" 
                                        value={this.state.title}
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <label htmlFor="price" id="priceLabel">
                                        {errorMessage ? errorMessage : 'Price'}
                                        <input
                                        type ="number" 
                                        id ="price" 
                                        name="price" 
                                        placeholder="price" 
                                        value={this.state.price}
                                        step=".01"
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <label htmlFor="description">
                                        Description
                                        <textarea
                                        id ="description" 
                                        name="description" 
                                        placeholder="Enter a Description" 
                                        value={this.state.description}
                                        onChange={this.handleInput}
                                        required  
                                        />
                                    </label>
                                    <button type="submit" disabled={errorMessage ? true : false} aria-disabled={errorMessage ? true : false}> Submit </button>
                                </fieldset>
                            </StyledForm>
                        )}
                        </Mutation> 
                    )}}
                </Query>
                </div>
            </StyledFormWrapper>
        );
    }
}

export default CreateItem;