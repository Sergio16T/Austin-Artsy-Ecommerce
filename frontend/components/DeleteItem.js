import React, { Component } from 'react';
import { Mutation } from 'react-apollo'; 
import gql from 'graphql-tag';
import Router from 'next/router'; 
import { ALL_ITEMS_QUERY } from './Items'; 
import { PAGINATION_QUERY } from './Pagination';

const DELETE_ITEM_MUTATION = gql`
    mutation DELETE_ITEM_MUTATION($id: ID!) {
        deleteItem(id: $id) {
            id
            title
        }
    }
`

let query = gql`
query ALL_ITEMS_QUERY($skip: Int = 0, $first: Int = 4) {
    items(skip: $skip, first: $first, orderBy: createdAt_ASC) {
        id
        title
        price
        description
        image 
        largeImage
    }
}
`; 

class DeleteItem extends Component {
    render() {
        return (
        <Mutation 
        mutation={DELETE_ITEM_MUTATION} 
        variables={{id: this.props.id}}
        refetchQueries={[
            {query: ALL_ITEMS_QUERY, variables:{skip: this.props.page * 4 -4}},
            {query: ALL_ITEMS_QUERY, variables:{skip: (this.props.page - 1) * 4 -4}},
            {query: ALL_ITEMS_QUERY, variables:{skip: (this.props.page + 1) * 4 -4}},
            {query: PAGINATION_QUERY },
        ]}
        >
            {(deleteItem, {error, loading}) => 
                <button onClick={async() => {
                    if(confirm("Are you sure you want to delete this item?")) { 
                    let res = await deleteItem().catch(err => {
                        alert(err.message); 
                    }); 
                    console.log('delete res', res); 
                    // Router.push({
                    //     pathname: "/items", 
                    //     query: {page: this.props.page}
                    // }); 
                    }
                }}>{this.props.children}</button>
            }
        </Mutation>
        
        );
    }
}

export default DeleteItem;