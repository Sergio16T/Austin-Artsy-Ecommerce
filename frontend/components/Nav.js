import React from 'react';
import Link from 'next/link'; 
import styled from 'styled-components'; 

const StyledNav = styled.ul`
        display: flex; 
        align-items: center; 
        justify-content: center;  
        .slidingBar {
            position: absolute; 
            height: 1px; 
            width: 0; 
            margin-top: 8px; 
            /* background: #0a0a36; */
            background: ${props => props.navBarColor ? '#0a0a36' : 'white' };
            left: 0;
            transition: width .4s; 
            transition-timing-function: ease;  
           
        }
        a {
            color: ${props => props.navBarColor ? '#0a0a36' : 'white' };
           /* color: white;  */
            font-weight: 600; 
            position: relative;  
            /* color: #0a0a36;  */
            top: 2px;
            transition: top .4s ease;
            transition: color .2s ease; 
            transition-delay: .4s; 
        }
        li {
            padding: 10px 0; 
            position:relative; 
            list-style: none; 
            margin: 0rem 1rem; 
            font-size: 1.4rem; 
            &:hover {
                cursor: pointer; 
                a{
                    top: -2px; 
                } 
                .slidingBar {
                    width: 100%;
                }
            }
        }

`;


const Nav = props => {
    return (
        <StyledNav navBarColor={props.navBarColor}>
            <Link href="/aboutus">
                <li>
                    <a>About Us</a>
                    <div className="slidingBar"></div>
                </li>
            </Link>
            <Link href="/items">
                <li>
                    <a>Shop</a>
                    <div className="slidingBar"></div>
                </li>
            </Link>
            <Link href="/sell">
                <li>
                    <a>Sell</a>
                    <div className="slidingBar"></div>
                </li>  
            </Link>
            <Link href="/orders">
                <li>
                    <a>Orders</a>
                    <div className="slidingBar"></div>
                </li>
            </Link>
            <Link href="/contact">
                <li>
                    <a>Contact</a>
                    <div className="slidingBar"></div>
                </li>
            </Link>
            <Link href="/signin">
                <li>
                    <a>Sign In</a>
                    <div className="slidingBar"></div>
                </li>
            </Link>
            <Link href="/signout">
                <li>
                    <a>Sign Out</a>
                    <div className="slidingBar"></div>
                </li>
            </Link>
            <SocialMediaIcon
            alt="linkedin"
            aria="linkedin"
            path="M30.07-10.9v-.076a.92.92 0 01-.05.077h.05zM3.977 17.558H.12V5.95h3.857zM2.047 4.364h-.024C.728 4.364-.11 3.474-.11 2.36-.11 1.22.755.353 2.075.353s2.132.866 2.158 2.005c0 1.113-.84 2.004-2.184 2.004zM17.89 17.558h-3.857v-6.21c0-1.56-.558-2.626-1.954-2.626-1.067 0-1.7.718-1.98 1.412-.103.247-.128.593-.128.94v6.484H6.114s.05-10.52 0-11.608H9.97v1.643c.514-.79 1.43-1.917 3.48-1.917 2.537 0 4.44 1.66 4.44 5.225v6.66z"
            navBarColor={props.navBarColor}
           />
            <SocialMediaIcon
            alt="instagram"
            aria="instagram"
            path="M127.999746,23.06353 C162.177385,23.06353 166.225393,23.1936027 179.722476,23.8094161 C192.20235,24.3789926 198.979853,26.4642218 203.490736,28.2166477 C209.464938,30.5386501 213.729395,33.3128586 218.208268,37.7917319 C222.687141,42.2706052 225.46135,46.5350617 227.782844,52.5092638 C229.535778,57.0201472 231.621007,63.7976504 232.190584,76.277016 C232.806397,89.7746075 232.93647,93.8226147 232.93647,128.000254 C232.93647,162.177893 232.806397,166.225901 232.190584,179.722984 C231.621007,192.202858 229.535778,198.980361 227.782844,203.491244 C225.46135,209.465446 222.687141,213.729903 218.208268,218.208776 C213.729395,222.687649 209.464938,225.461858 203.490736,227.783352 C198.979853,229.536286 192.20235,231.621516 179.722476,232.191092 C166.227425,232.806905 162.179418,232.936978 127.999746,232.936978 C93.8200742,232.936978 89.772067,232.806905 76.277016,232.191092 C63.7971424,231.621516 57.0196391,229.536286 52.5092638,227.783352 C46.5345536,225.461858 42.2700971,222.687649 37.7912238,218.208776 C33.3123505,213.729903 30.538142,209.465446 28.2166477,203.491244 C26.4637138,198.980361 24.3784845,192.202858 23.808908,179.723492 C23.1930946,166.225901 23.0630219,162.177893 23.0630219,128.000254 C23.0630219,93.8226147 23.1930946,89.7746075 23.808908,76.2775241 C24.3784845,63.7976504 26.4637138,57.0201472 28.2166477,52.5092638 C30.538142,46.5350617 33.3123505,42.2706052 37.7912238,37.7917319 C42.2700971,33.3128586 46.5345536,30.5386501 52.5092638,28.2166477 C57.0196391,26.4642218 63.7971424,24.3789926 76.2765079,23.8094161 C89.7740994,23.1936027 93.8221066,23.06353 127.999746,23.06353 M127.999746,0 C93.2367791,0 88.8783247,0.147348072 75.2257637,0.770274749 C61.601148,1.39218523 52.2968794,3.55566141 44.1546281,6.72008828 C35.7374966,9.99121548 28.5992446,14.3679613 21.4833489,21.483857 C14.3674532,28.5997527 9.99070739,35.7380046 6.71958019,44.1551362 C3.55515331,52.2973875 1.39167714,61.6016561 0.769766653,75.2262718 C0.146839975,88.8783247 0,93.2372872 0,128.000254 C0,162.763221 0.146839975,167.122183 0.769766653,180.774236 C1.39167714,194.398852 3.55515331,203.703121 6.71958019,211.845372 C9.99070739,220.261995 14.3674532,227.400755 21.4833489,234.516651 C28.5992446,241.632547 35.7374966,246.009293 44.1546281,249.28042 C52.2968794,252.444847 61.601148,254.608323 75.2257637,255.230233 C88.8783247,255.85316 93.2367791,256 127.999746,256 C162.762713,256 167.121675,255.85316 180.773728,255.230233 C194.398344,254.608323 203.702613,252.444847 211.844864,249.28042 C220.261995,246.009293 227.400247,241.632547 234.516143,234.516651 C241.632039,227.400755 246.008785,220.262503 249.279912,211.845372 C252.444339,203.703121 254.607815,194.398852 255.229725,180.774236 C255.852652,167.122183 256,162.763221 256,128.000254 C256,93.2372872 255.852652,88.8783247 255.229725,75.2262718 C254.607815,61.6016561 252.444339,52.2973875 249.279912,44.1551362 C246.008785,35.7380046 241.632039,28.5997527 234.516143,21.483857 C227.400247,14.3679613 220.261995,9.99121548 211.844864,6.72008828 C203.702613,3.55566141 194.398344,1.39218523 180.773728,0.770274749 C167.121675,0.147348072 162.762713,0 127.999746,0 Z M127.999746,62.2703115 C91.698262,62.2703115 62.2698034,91.69877 62.2698034,128.000254 C62.2698034,164.301738 91.698262,193.730197 127.999746,193.730197 C164.30123,193.730197 193.729689,164.301738 193.729689,128.000254 C193.729689,91.69877 164.30123,62.2703115 127.999746,62.2703115 Z M127.999746,170.667175 C104.435741,170.667175 85.3328252,151.564259 85.3328252,128.000254 C85.3328252,104.436249 104.435741,85.3333333 127.999746,85.3333333 C151.563751,85.3333333 170.666667,104.436249 170.666667,128.000254 C170.666667,151.564259 151.563751,170.667175 127.999746,170.667175 Z M211.686338,59.6734287 C211.686338,68.1566129 204.809755,75.0337031 196.326571,75.0337031 C187.843387,75.0337031 180.966297,68.1566129 180.966297,59.6734287 C180.966297,51.1902445 187.843387,44.3136624 196.326571,44.3136624 C204.809755,44.3136624 211.686338,51.1902445 211.686338,59.6734287 Z"
            viewBox="0 0 256 256"
            navBarColor={props.navBarColor}
            />
            <SocialMediaIcon
            alt="twitter"
            aria="twitter"
            path="M18 3.418a7.332 7.332 0 01-2.12.58 3.7 3.7 0 001.623-2.042 7.413 7.413 0 01-2.345.896 3.693 3.693 0 00-2.696-1.166A3.692 3.692 0 008.865 6.22a10.485 10.485 0 01-7.612-3.86 3.692 3.692 0 001.143 4.93 3.687 3.687 0 01-1.673-.46v.047c0 1.79 1.273 3.28 2.963 3.62a3.673 3.673 0 01-1.668.064 3.697 3.697 0 003.45 2.565 7.416 7.416 0 01-4.587 1.58c-.296 0-.59-.017-.88-.05a10.46 10.46 0 005.66 1.658c6.795 0 10.51-5.627 10.51-10.507 0-.16-.004-.32-.01-.478A7.526 7.526 0 0018 3.416z"
            navBarColor={props.navBarColor}
            />
            <SocialMediaIcon
            alt="github"
            aria="github"
            path="M12.11 10.28c-.702 0-1.272.79-1.272 1.76 0 .974.57 1.762 1.272 1.762.702 0 1.27-.788 1.27-1.76.002-.973-.568-1.76-1.27-1.76zm4.428-4.58c.145-.358.152-2.387-.62-4.33 0 0-1.772.194-4.453 2.034C10.903 3.248 9.95 3.17 9 3.17s-1.902.078-2.465.233C3.855 1.563 2.083 1.37 2.083 1.37c-.773 1.943-.766 3.972-.62 4.33C.553 6.685 0 7.87 0 9.486c0 7.03 5.832 7.143 7.304 7.143h3.392c1.472 0 7.304-.115 7.304-7.145 0-1.617-.554-2.8-1.462-3.786zm-7.51 10.064h-.055c-3.69 0-6.564-.44-6.564-4.026 0-.86.302-1.656 1.022-2.318 1.2-1.103 3.233-.52 5.54-.52h.055c2.306 0 4.34-.582 5.54.52.72.662 1.023 1.46 1.023 2.318 0 3.586-2.873 4.026-6.562 4.026zM5.89 10.28c-.702 0-1.27.79-1.27 1.76 0 .974.568 1.762 1.27 1.762.703 0 1.272-.788 1.272-1.76 0-.973-.57-1.76-1.272-1.76z"
            navBarColor={props.navBarColor}
            />
        </StyledNav>
    );
};


const SocialMediaIcon = (props) => {
    return (
        <li className="socialListItem">
        <a 
        alt={props.alt} 
        aria-label={props.aria}
        href="/" 
        rel="noopener" 
        target="_blank"> 
        {props.viewBox ? (
        <svg 
        className="social-media-list__icon" 
        xmlns="http://www.w3.org/2000/svg" 
        height="18" 
        width="18"
        viewBox={props.viewBox}
        >
        <path fill="currentColor"/*{props.navBarColor ? '#0a0a36' : 'white'}*/ d={props.path}></path>
        </svg> 
        )
        : ( 
        <svg 
        className="social-media-list__icon" 
        xmlns="http://www.w3.org/2000/svg" 
        height="18" 
        width="18">
        <path fill="currentColor"/*{props.navBarColor ? '#0a0a36' : 'white'}*/ d={props.path}></path>
        </svg> 
        ) }
        </a>
        </li>
    ); 
}

export default Nav;