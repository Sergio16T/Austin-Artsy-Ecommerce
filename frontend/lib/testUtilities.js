import casual from 'casual'; 
//.seed() is to provide ability to use a specific seed in order to get a repeatable random sequence for our testing purposes
casual.seed(111); 

const mockItem = () => ({
    // __typename: 'Item',
    id: '123',
    price: 5000,
    user: null,
    image: 'dog.jpg',
    title: 'dogs are the best',
    description: 'dogs description',
    largeImage: 'dog-large.jpg',
}); 

const mockUserWithItemsInCart = () =>  ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ["ADMIN"], 
    cart: [{
        id: 'abc', 
        quantity: 2, 
        item: {
            id: 'dgh', 
            price: 15000, 
            image: 'www.dog.image.com',
            title: "puppy"
        }
        },
    ]
}); 
const mockUser = () => ({
    __typename: 'User', 
    id: "123", 
    name: casual.name, 
    email: casual.email, 
    permissions: ['ADMIN'], 
    cart: [],
    orders: []
}); 
const mockOrderItem = () => ({
    __typename: 'OrderItem',
    id: casual.uuid,
    image: `${casual.word}.jpg`,
    title: casual.words(),
    price: 4234,
    quantity: 1,
    description: casual.words(),
  });
  
  const mockOrder = () => ({
    __typename: 'Order',
    id: 'ord123',
    charge: 'ch_123',
    total: 40000,
    items: [mockOrderItem(), mockOrderItem()],
    createdAt: '2018-04 - 06T19: 24: 16.000Z',
    user: mockUser(),
  });
  
  const mockCartItem = overrides => ({
    __typename: 'CartItem',
    id: 'omg123',
    quantity: 3,
    item: mockItem(),
    user: mockUser(),
    ...overrides,
  });
//mock LocalStorage 
class MockLocalStorage {
    constructor(){
        this.store={}; 
    }

    clear() {
        this.store = {}; 
    }
    getItem(key) {
        return this.store[key] || null; 
    }
    setItem(key, value) {
        this.store[key] = value.toString(); 
    }
    removeItem(key) {
        delete this.store[key]; 
    }
}

export { 
    mockItem, 
    MockLocalStorage, 
    mockUser, 
    mockCartItem, 
    mockOrderItem, 
    mockOrder }; 