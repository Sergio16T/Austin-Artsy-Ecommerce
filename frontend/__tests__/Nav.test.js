import { render, screen, act } from '@testing-library/react'; 
import { CURRENT_USER_QUERY } from '../components/User'; 
import Nav from '../components/Nav'; 
import { MockedProvider } from '@apollo/react-testing'; 
import { mockUser, mockCartItem } from '../lib/testUtilities'; 
import wait from 'waait'; 

const notSignedInMocks = [{ 
    request: { query: CURRENT_USER_QUERY}, 
    result: { data: { user: null } }, 
}, 
]; 

const signedInMocks = [{
request: { query: CURRENT_USER_QUERY }, 
result: { data: { user: mockUser() } }, 
}];

const signedInMocksWithCart=[{
    request: { query: CURRENT_USER_QUERY }, 
    result: { data: { user: {
        ...mockUser(), 
        cart: [mockCartItem(), mockCartItem(), mockCartItem() ]
        }, 
    	},
    },
}];

jest.mock("next/router", () => ({
    useRouter() {
      return {
        route: "",
        pathname: "",
        query: "",
        asPath: "",
      };
    },
  }));
//spyOn Creates a mock function similar to jest.fn but also tracks calls to object[methodName]. Returns a Jest mock function.
// creating a mockfunction of the mock I declared globally above for useRouter. This will track calls to useRouter
const useRouter = jest.spyOn(require("next/router"), "useRouter"); 

describe('<Nav/>', ()=> {
    it('renders a minimal nav when signed out ', async () => {
        // mockImplementation Accepts a function that should be used as the implementation of the mock.. mockImplemetation used to overwrite the original function when used with the useRouter value which is initalized with jest.spyOn() 
        useRouter.mockImplementation(() => ({
            route: "/",
            pathname: "/",
            query: "",
            asPath: "",
          }));
        const { container } = render(
            <MockedProvider mocks={notSignedInMocks}>
                    <Nav/>
            </MockedProvider>
        ); 
        await act( async() => {
            await wait()
        }); 
       
        const loginLink= screen.getByTestId('login'); 
        // const contactLink = screen.getByTestId('contact'); 
        const anchors = container.querySelectorAll('a'); 
        expect(anchors).toHaveLength(5); 
        expect(loginLink).toBeInTheDocument(); 
        // expect(contactLink).toBeInTheDocument(); 
        expect(container).toMatchSnapshot(); 
    }); 
    it('renders a more robust nav when signed in', async() => {
        useRouter.mockImplementation(() => ({
            route: "/",
            pathname: "/",
            query: "",
            asPath: "",
          }));
        const { container } = render(
            <MockedProvider mocks={signedInMocks}>
                    <Nav/>
            </MockedProvider>
        ); 
        await act( async() => {
            await wait()
        });
        const anchors = container.querySelectorAll('a'); 
        expect(anchors).toHaveLength(8); 
        expect(container).toMatchSnapshot(); 
        expect(screen.getByText('Shop')).toBeInTheDocument(); 
    }); 
    it('renders with cart items' , async () => {
        useRouter.mockImplementation(() => ({
            route: "/items",
            pathname: "/items",
            query: "",
            asPath: "",
          }));
        const { container, debug } = render(
            <MockedProvider mocks={signedInMocksWithCart}>
                    <Nav/>
            </MockedProvider>
        ); 
        await act( async () => {
            await wait(); 
        }); 
        // debug()
        const cartCount = screen.getByTestId('cartCount'); 
        expect(cartCount.innerHTML).toEqual("9"); 
    });
}); 