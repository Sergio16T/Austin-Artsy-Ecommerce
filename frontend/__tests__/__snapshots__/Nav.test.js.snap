// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav/> renders a minimal nav when signed out  1`] = `
Array [
  <Link
    href="/contact"
  >
    <li
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        Contact
      </a>
      <div
        className="slidingBar"
        id="/contact"
      />
    </li>
  </Link>,
  <Link
    href="/signup"
  >
    <li
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        Log In
      </a>
      <div
        className="slidingBar"
        id="/signup"
      />
    </li>
  </Link>,
]
`;

exports[`<Nav/> renders a more robust nav when signed in 1`] = `
Array [
  <Link
    href="/items"
  >
    <li
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        Shop
      </a>
      <div
        className="slidingBar"
        id="/items"
      />
    </li>
  </Link>,
  <Link
    href="/sell"
  >
    <li
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        Sell
      </a>
      <div
        className="slidingBar"
        id="/sell"
      />
    </li>
  </Link>,
  <Link
    href="/orders"
  >
    <li
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        Orders
      </a>
      <div
        className="slidingBar"
        id="/orders"
      />
    </li>
  </Link>,
  <Link
    href="/contact"
  >
    <li
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <a>
        Contact
      </a>
      <div
        className="slidingBar"
        id="/contact"
      />
    </li>
  </Link>,
]
`;

exports[`<Nav/> renders with cartItems 1`] = `
<span
  id="cartCount"
>
  9
</span>
`;
